Add DNS oracle protocol

Jim8y opened the pull request:
## Summary
- add DNS-over-HTTPS oracle protocol and config wiring
- document dns:// usage and add tests for certificate parsing

## Testing
- not run (not requested)

Jim8y committed:
Add DNS oracle protocol

shargon reviewed (commented):
I think that it's better to define the dns server, and make a dns query, instead of http query

erikzhang commented:
> I think that it's better to define the dns server, and make a dns query, instead of http query

We need DNS over HTTPS.

erikzhang commented on docs/oracle-dns-protocol.md:
````diff
@@ -0,0 +1,128 @@
+## Constructing dns:// URLs
+
+```
+dns://<base-domain>/<path-selector>?selector=<label>&name=<override>&type=<rr-type>&format=<output>
````
We need to be consistent with [RFC4501](https://www.rfc-editor.org/rfc/rfc4501.html).

Jim8y committed:
Align oracle DNS protocol with RFC 4501

Jim8y committed:
Guard DoH certificate parsing against oversized payloads

shargon commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,503 @@
+        CertificatePublicKey key = new()
+        {
+            Algorithm = cert.PublicKey.Oid?.FriendlyName ?? cert.PublicKey.Oid?.Value,
+            Encoded = Convert.ToBase64String(cert.GetPublicKey())
```
I think that it's better to construct this object inside the try/catch clause, it's possible to throw [CryptographicException](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptographicexception?view=net-9.0) during cert.GetPublicKey()

Jim8y replied:
This path no longer constructs an X509Certificate or calls cert.GetPublicKey(). CERT RDATA is returned as raw RFC 4398 data, so public key extraction cannot throw here anymore.

erikzhang commented on docs/oracle-dns-protocol.md:
```diff
@@ -0,0 +1,129 @@
+
+- `domain` is the DNS owner name (relative or absolute). Percent-encoding and escaped dots (`%5c.`) follow RFC 4501 rules.
+- `domain` must not include additional path segments; only the owner name belongs here.
+- `authority` is the optional DNS server hint from RFC 4501. The oracle still uses the DoH resolver configured in `OracleService.json`.
```
Why not use the user-specified authority?

Jim8y replied:
is that necessary?  can be updated if you think its a right thing to do. for me, current design is sufficient.

erikzhang replied:
Relying on a fixed DNS server is unreliable and it may fail. At that time, we must modify the configuration file before we can repair it. We can completely use the server specified by the user in the URI for querying.

erikzhang commented on docs/oracle-dns-protocol.md:
```diff
@@ -0,0 +1,129 @@
+- `dns:1alhai._domainkey.icloud.com?TYPE=TXT` — DKIM TXT record.
+- `dns:simon.example.org?TYPE=CERT;FORMAT=x509` — extract the X.509 payload into `Certificate`.
+- `dns://192.168.1.1/ftp.example.org?TYPE=A` — RFC-compliant authority form (authority is ignored; the configured DoH endpoint is used).
+- `dns:ignored?name=weird%5c.label.example&type=TXT` — uses the `name` override (decoded to `weird\.label.example`).
```
Why can a DNS name contain backslashes (`\`)?

erikzhang committed:
Merge branch 'master' into feature/dns-doh

Wi1l-B0t commented on docs/oracle-dns-protocol.md:
```diff
@@ -0,0 +1,129 @@
+    // ...
+    "Dns": {
+      "EndPoint": "https://cloudflare-dns.com/dns-query",
+      "Timeout": 5000
```
What Unit ?

Millseconds?

```suggestion
      "TimeoutMilliseconds": 5000
```

ajara87 committed:
Merge branch 'master' into feature/dns-doh

doubiliu commented on plugins/OracleService/OracleSettings.cs:
```diff
@@ -36,6 +36,19 @@ public NeoFSSettings(IConfigurationSection section)
+
+    public DnsSettings(IConfigurationSection section)
+    {
+        string endpoint = section.GetValue("EndPoint", "https://cloudflare-dns.com/dns-query");
```
The current investigation has not found it. DNSSEC is regarded as a paid item by large email service providers and is not universal. In addition, I think the impact of the previous world-class disaster was limited, because our biggest impact was the suspension of services, and the problem of financial security was not too big.

ajara87 committed:
Merge branch 'master' into feature/dns-doh

ajara87 committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
Update plugins/OracleService/OracleSettings.cs

Co-authored-by: Will <201105916+Wi1l-B0t@users.noreply.github.com>

Jim8y committed:
Update docs/oracle-dns-protocol.md

Co-authored-by: Will <201105916+Wi1l-B0t@users.noreply.github.com>

Jim8y committed:
Update DNS oracle to RFC 8484 application/dns-message format

- Replace application/dns-json with standard application/dns-message
- Implement DNS wire format (RFC 1035) for query/response encoding
- Use HTTP POST method per RFC 8484 specification
- Add DNS name compression pointer support
- Support user-specified authority in URI (dns://resolver/domain)
- Fix CryptographicException handling in BuildPublicKey
- Move Accept header to constructor
- Add comprehensive unit tests for wire format handling
- Add integration tests for Cloudflare, Google, and Quad9 DoH endpoints
- Update documentation with RFC 8484 compliance details

Jim8y commented:
- ✅ application/dns-json → application/dns-message (RFC 8484)
  - ✅ HTTP GET → HTTP POST
  - ✅ JSON parsing → DNS wire format (RFC 1035)
  - ✅ Support user-specified authority (dns://resolver/domain)
  - ✅ Fix CryptographicException handling in BuildPublicKey
  - ✅ Move Accept header to constructor
  - ✅ Integration tests covering Cloudflare, Google, Quad9

Jim8y committed:
Simplify DNS DoH handling and tighten safeguards

Jim8y committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
Align oracle DNS protocol with RFC4501

ajara87 committed:
Merge branch 'master' into feature/dns-doh

erikzhang commented on plugins/OracleService/OracleService.json:
```diff
@@ -14,6 +14,10 @@
     },
+    "Dns": {
+      "EndPoint": "https://cloudflare-dns.com/dns-query",
+      "TimeoutMilliseconds": 5000
```
L15 and L19, one is `Timeout` and the other is `TimeoutMilliseconds`. It's best to keep them consistent.

Jim8y replied:
The configuration now uses Timeout consistently for Https, NeoFS, and Dns, and DnsSettings reads Timeout directly.

Jim8y committed:
Oracle DNS: return struct result for contracts

Jim8y committed:
Merge branch 'master' into feature/dns-doh

shargon commented on plugins/OracleService/OracleService.cs:
```diff
@@ -528,6 +553,35 @@ public static byte[] Filter(string input, string filterArgs)
+        return data.Length <= maxLen ? data : data[..maxLen] + "...";
+    }
+
+    internal static bool TryDecodeDnsStackItemPayload(string payload, out byte[] result)
```
```suggestion
    internal static bool TryDecodeDnsStackItemPayload(string payload, [NotNullWhen(true)] out byte[]? result)
```

Jim8y committed:
Update plugins/OracleService/OracleService.cs

Co-authored-by: Christopher Schuchardt <8141309+cschuchardt88@users.noreply.github.com>

Jim8y committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
style: format oracle dns files

Jim8y committed:
fix: restore compatibility with current Neo package

Jim8y committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
Use PipeReader for DoH responses

Jim8y commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -250,14 +221,45 @@ private async Task<DnsMessage> ResolveAsync(string name, ushort type, Uri resolv
+        if (response.Content.Headers.ContentLength.HasValue && response.Content.Headers.ContentLength > OracleResponse.MaxResultSize)
+            throw new ResponseTooLargeException();
+
+        using Stream stream = await response.Content.ReadAsStreamAsync(cancellation);
```
Updated the response reader to use PipeReader while preserving the OracleResponse.MaxResultSize bound.

shargon committed:
check bool result

shargon commented on tests/Neo.Plugins.OracleService.Tests/UT_OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,1201 @@
+            OracleDnsProtocol.BuildQueryName(uri);
+            Assert.Fail("Expected FormatException for missing dnsname.");
+        }
+        catch (FormatException)
```
Could you use `Assert.ThrowsExactly`? this pattern it's dirty

shargon commented on plugins/OracleService/OracleService.cs:
```diff
@@ -542,6 +569,41 @@ public static byte[] Filter(string input, string? filterArgs)
+            result = Convert.FromBase64String(payload);
+            return true;
+        }
+        catch (Exception ex) when (ex is ArgumentNullException or FormatException)
```
```suggestion
        catch (Exception ex)
```
Avoid unexpected errors that can deal in DoS

Jim8y committed:
Address Oracle DNS review comments

Jim8y committed:
Address DNS oracle review feedback

Jim8y commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,711 @@
+
+        if (dnsResponse is null)
+            return (OracleResponseCode.Error, "Invalid DNS response.");
+
```
Handled in 09d4a4b. The unreachable null response branch was removed.

Jim8y commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,711 @@
+        message.Add(0x00);
+        message.Add(0x01);
+
+        return [.. message];
+    }
+
+    /// <summary>
+    /// Encodes a domain name in DNS wire format (RFC 1035 section 4.1.2).
+    /// </summary>
+    private static void EncodeDnsName(List<byte> buffer, string name)
+    {
+        if (string.IsNullOrEmpty(name) || name == ".")
+        {
+            buffer.Add(0x00);
+            return;
+        }
+
+        string[] labels = name.TrimEnd('.').Split('.');
+        foreach (string label in labels)
+        {
+            if (label.Length > 63)
+                throw new FormatException($"DNS label exceeds 63 characters: {label}");
+            byte[] labelBytes = Encoding.ASCII.GetBytes(label);
+            buffer.Add((byte)labelBytes.Length);
+            buffer.AddRange(labelBytes);
+        }
+        buffer.Add(0x00); // Root label
+    }
```
Handled in 09d4a4b. DNS name encoding now enforces the 255-octet wire-format limit and has coverage for the maximum valid name and an overlong name.

Jim8y commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,711 @@
+        }
+        finally
+        {
+            await reader.CompleteAsync();
+        }
+    }
+
+    private async Task EnsureEndpointAllowed(Uri resolverEndpoint, CancellationToken cancellation)
+    {
+        if (OracleSettings.Default.AllowPrivateHost)
+            return;
+
+        if (IsPrivateEndpoint(resolverEndpoint.Host))
+            throw new InvalidOperationException("Private resolver endpoints are not allowed.");
+
+        try
+        {
+            IPHostEntry entry = await Dns.GetHostEntryAsync(resolverEndpoint.Host, cancellation);
+            if (entry.IsInternal())
+                throw new InvalidOperationException("Private resolver endpoints are not allowed.");
+        }
+        catch (SocketException ex)
+        {
+            throw new InvalidOperationException($"Failed to resolve resolver endpoint: {ex.Message}", ex);
+        }
+    }
+
+    private static bool IsPrivateEndpoint(string host)
+    {
+        if (string.IsNullOrWhiteSpace(host))
+            return false;
+
+        if (host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+            return true;
+
+        if (IPAddress.TryParse(host, out IPAddress? address))
+            return address.IsInternal();
+
+        return false;
+    }
```
Handled in 09d4a4b. Localhost resolver suffixes are rejected before resolver lookup, with coverage for sub.localhost.

Jim8y replied:
Follow-up in 4c4583f. The default DoH HTTP handler now resolves the resolver host during connection setup, rejects the connection if any resolved address is private, and opens the socket directly to the validated address instead of letting HttpClient resolve it again. I also added coverage for mixed public/private resolved address lists.

Jim8y commented on plugins/OracleService/OracleService.cs:
```diff
@@ -301,8 +303,20 @@ private async Task ProcessRequestAsync(DataCache snapshot, OracleRequest req)
+                code = OracleResponseCode.Error;
+                PluginLogger?.Warning("Invalid DNS stack item payload: {OriginalTxid}", req.OriginalTxid);
+            }
+        }
```
Handled in 09d4a4b. The final response log is emitted after DNS stack payload validation, so the logged response code matches any Error conversion from decode failure.

Jim8y committed:
Pin DNS resolver connections

shargon committed:
Merge branch 'master' into feature/dns-doh

Jim8y committed:
Harden DNS oracle response handling

ajara87 committed:
Merge branch 'master' into feature/dns-doh

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+        if (rdata is null || rdata.Length == 0)
+            return string.Empty;
+
+        return type switch
```
**[bug]** `FormatRData` only special-cases A/AAAA/TXT/CERT. CNAME, NS, MX, SRV, and SOA (all in `RecordTypeLookup`) fall through to `Convert.ToBase64String(rdata)` of the raw wire bytes. Those RDATA fields contain domain names that almost always use RFC 1035 compression pointers (`0xC0 0x0C` pointing at the question name is typical). The oracle payload then contains the pointer bytes, not the hostname, so a CNAME answer is useless to contracts. Two resolvers (or the same resolver with/without compression) serialize different bytes for the same record, so oracle nodes will not agree. `SkipDnsName` / `DecodeDnsName` already exist but are not applied to RDATA.

**Suggestion:** Expand name-valued RDATA with `DecodeDnsName` (and parse MX/SRV/SOA fields) into a stable text form, or drop those types until that is implemented. Never copy compressed wire bytes into the consensus payload.

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+        ResultAnswer[] answers;
+        try
+        {
+            answers = dnsResponse.Answers
```
**[bug]** Every resource record in the DNS answer section is serialized, not just the requested type, and owner names keep wire-format case (`a.Name.TrimEnd('.')` at line 189). Sorting is case-insensitive, so order can match while the ByteStrings differ (`WWW.Example.COM` vs `www.example.com`). CNAME chains make this worse: a TYPE=A lookup often returns CNAME+A, while another DoH server returns only A. Docs claim answers are normalized so all oracle nodes serialize the same payload; TTL=0 plus sort is not enough. Disagreeing payloads fail the request with `ConsensusUnreachable` rather than forking, but `dns:` will be unreliable in production.

**Suggestion:** Keep only records whose type equals the query type (or document and implement a fixed CNAME-follow policy). Lowercase names before sort and serialize. Deduplicate `(Name, Type, Data)`. Document that `dns:` URLs without an authority require every oracle operator to use the same `Dns.EndPoint`.

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+            return;
+        }
+
+        string[] labels = name.TrimEnd('.').Split('.');
```
**[bug]** RFC 4501 (and this PR’s docs) encode a dot inside a label as `\.`, written in the URI as `%5c.`. `NormalizeLabel` unescapes `%5c` to `\`, then `EncodeDnsName` does `name.TrimEnd('.').Split('.')`, so `exa%5c.mple.com` becomes labels `exa\` and `mple` instead of one label `exa.mple`. RFC 1035 `\DDD` octal escapes are also ignored. The documented example form does not round-trip.

**Suggestion:** Split labels on unescaped dots only, and interpret `\.` / `\DDD` per RFC 1035 master-file escaping before encoding the query.

cschuchardt88 commented on plugins/OracleService/OracleService.cs:
```diff
@@ -312,7 +326,20 @@ private async Task ProcessRequestAsync(DataCache snapshot, OracleRequest req)
-                    result = Filter(data!, request.Filter);
+                    if (dnsStackOutput)
+                    {
+                        if (!string.IsNullOrEmpty(request.Filter))
```
**[bug]** `ProcessRequestAsync` loops all oracle requests that share a URL and the catch sets the shared `code` on filter failure. Any non-empty filter on a `dns:` request throws (`Filter is not supported for dns: requests.`). If one contract uses a filter and another uses the same URL with an empty filter, iteration order decides whether the valid request is also answered as `Error`. HTTPS had a similar catch, but DNS now throws on every non-empty filter, so poisoning is the common case rather than a bad JSONPath.

**Suggestion:** Use a per-iteration `responseCode` (and result). Do not assign `code = OracleResponseCode.Error` for a sibling request. Rejecting a DNS filter should not change other requests for that URL.

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+
+    internal static IPAddress[] ValidateEndpointAddresses(IPAddress[] addresses)
+    {
+        if (!OracleSettings.Default.AllowPrivateHost && addresses.Any(p => p.IsInternal()))
```
**[bug]** User-specified `dns://authority/` makes every oracle node POST to `https://{authority}/dns-query`. Private-host checks call `IPAddress.IsInternal()`, which only treats 10/8, 127/8, 172.16/12, 192.168/16, loopback, and unspecified as internal. Link-local `169.254.0.0/16` (cloud metadata), CGNAT `100.64.0.0/10`, IPv6 ULA `fc00::/7`, and IPv6 link-local `fe80::/10` pass. `dns://169.254.169.254/example.com?TYPE=A` therefore reaches `https://169.254.169.254/dns-query`. HTTPS has the same helper gap; this PR reuses it in a new user-controlled resolver path and ConnectCallback. HTTPS maps private hosts to `Forbidden`; this path throws and becomes `Error`.

**Suggestion:** Treat 169.254/16, 100.64/10, IPv6 ULA, and IPv6 link-local as internal in the resolver allow-check (or extend `Helper.IsInternal` and use it here). Return `OracleResponseCode.Forbidden` like `OracleHttpsProtocol`. Keep blocking mixed public+private A/AAAA sets.

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+                throw new FormatException("DNS name contains an empty label.");
+            if (label.Length > 63)
+                throw new FormatException($"DNS label exceeds 63 characters: {label}");
+            byte[] labelBytes = Encoding.ASCII.GetBytes(label);
```
**[suggestion]** `Encoding.ASCII.GetBytes(label)` uses the default replacement fallback, so a percent-decoded IDN label (`exämple`) is queried as `ex?mple` with no error. RFC 4501 treats `dnsname` as an IDN-unaware slot; non-ASCII should be rejected (or converted to punycode), not silently corrupted.

**Suggestion:** Use `EncoderExceptionFallback` or reject any label with non-ASCII octets. If IDN is desired, apply IDNA ToASCII first.

cschuchardt88 commented on plugins/OracleService/OracleSettings.cs:
```diff
@@ -36,6 +36,19 @@ public NeoFSSettings(IConfigurationSection section)
+    public DnsSettings(IConfigurationSection section)
+    {
+        string endpoint = section.GetValue("EndPoint", "https://cloudflare-dns.com/dns-query");
+        EndPoint = new Uri(endpoint, UriKind.Absolute);
```
**[suggestion]** `DnsSettings` accepts any absolute URI as `EndPoint` (http, ftp, custom path). Authority-form URLs are forced to `https` + `/dns-query`, but the configured fallback is not. Plaintext DoH and a per-node Cloudflare default both undermine the “identical results” goal when the URI omits authority. Geo-DNS A records will still differ even with one provider.

**Suggestion:** Require `https` for `Dns.EndPoint`. Document that operators must share the same endpoint, or prefer authority-form `dns://` URLs for anything that must reach consensus.

cschuchardt88 commented on plugins/OracleService/Protocols/OracleDnsProtocol.cs:
```diff
@@ -0,0 +1,807 @@
+    /// </summary>
+    private Uri GetResolverEndpoint(Uri uri)
+    {
+        // Check if URI has an authority (e.g., dns://resolver.example.com/domain)
```
**[suggestion]** Several comments restate the next line (`Check if URI has an authority`, `Fall back to configured endpoint`, header/flag byte comments in `BuildDnsQuery`, `Parse answer section`). `SkipDnsName` is unused dead code left from an earlier parser. Comments should explain non-obvious constraints (why redirects are disabled, why TTL is zeroed), not narrate the wire layout.

**Suggestion:** Delete `SkipDnsName` and the WHAT comments. Keep a short note only where the RFC constraint is not obvious from the code.
